guides/customizing_saving_and_serialization.py

```diff
@@ -0,0 +1,334 @@
+additional information that your model needs.
+
+For example, NLP domain layers such as TextVectorization layers and IndexLookup layers
+may need to store their associated vocabulary (or lookup table) in a text file upon
+saving.
+
+Let's take at the basics of this workflow with a simple file `assets.txt`.
+
+**Example:**
+"""
+
+
+@keras.saving.register_keras_serializable(package="my_custom_package")
+class LayerWithCustomAssets(keras.layers.Dense):
+    def build(self, input_shape):
+        if not hasattr(self, "assets"):
+            self.assets = "Mary had a <unk> lamb."
+        return super().build(input_shape)
+
+    def save_assets(self, inner_path):
+        with open(os.path.join(inner_path, "assets.txt"), "w") as f:
+            f.write(
+                self.assets
+            )  # Writes the assets (sentence) to text file at save time.
+
+    def load_assets(self, inner_path):
+        with open(os.path.join(inner_path, "assets.txt"), "r") as f:
+            text = f.read()  # Reads the assets (sentence) from text file at load time.
+        self.assets = text.replace("<unk>", "little")
+
+
+@keras.saving.register_keras_serializable(package="my_custom_package")
+class ModelWithCustomAssets(keras.Model):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.custom_dense = LayerWithCustomAssets(1)
+
+    def call(self, inputs):
+        return self.custom_dense(inputs)
+
+
+model = ModelWithCustomAssets()
+
+x = np.random.random((10, 10))
+y = model(x)
+
+model.save("custom_assets_model.keras")
+restored_model = keras.models.load_model("custom_assets_model.keras")
+
+np.testing.assert_string_equal(
+    restored_model.custom_dense.assets, "Mary had a little lamb."
+)
+
+"""
+## `build` and `compile` saving customization
+"""
+
+"""
+### `get_build_config()` and `build_from_config()`
+
+These methods work together to save the layer's built states and restore them upon
+loading.
+
+By default, this only includes a build config dictionary with the layer's input shape,
+but overriding these methods can be used to include further Variables and Lookup Tables
+that can be useful to restore for your built model.
+
+**Example:**
+"""
+
+
+@keras.saving.register_keras_serializable(package="my_custom_package")
+class LayerWithCustomBuild(keras.layers.Layer):
+    def __init__(self, units=32, **kwargs):
+        super().__init__(**kwargs)
+        self.units = units
+
+    def call(self, inputs):
+        return tf.matmul(inputs, self.w) + self.b
+
+    def get_config(self):
+        return dict(units=self.units, **super().get_config())
+
+    def build(self, input_shape, layer_init):
+        # Note the customization in overriding `build()` adds an extra argument.
+        # Therefore, we will need to manually call build with `layer_init` argument
+        # before the first execution of `call()`.
+        super().build(input_shape)
```

Comment:
Is the super() call necessary?

Reply:
Great question! Here it's necessary because our overriding `build()` only adds an extra weight and bias that depends on a specified initializer arg. Building the weights according to the input shape is still needed, which is why we call the base behavior with `super().build(input_shape)`.